### DISCUSSION TITLE
lookup docker.io - error msg: routines:ssl3_get_record:wrong version number

### DISCUSSION THREAD

**Discussion by rcarre** [renovatebot/renovate#24136] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.68.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

No way to lookup for dependencies into docker.io ... also tried `"insecureRegistry": true` with no success
Thanks for your help

**Config.js**
```
DEBUG: Combined config
       "config": {
         "platform": "gitlab",
         "dependencyDashboardAutoclose": true,
         "repositoryCache": "enabled",
         "autodiscover": true,
         "extends": ["mergeConfidence:all-badges"],
         "binarySource": "global",
         "ignorePrAuthor": true,
         "onboardingConfig": {
           "extends": [
             "config:recommended",
             ":disableDependencyDashboard",
             ":semanticCommitTypeAll(fix)"
           ],
           "separateMajorMinor": false
         },
         "allowedPostUpgradeCommands": ["./.renovate-post-upgrade"],
         "hostRules": [
           {
             "hostType": "docker",
             "matchHost": "docker.io",
             "username": "*********",
             "password": "***********"
           },
           {
             "matchHost": "registry.*****.orange",
             "username": "***********",
             "password": "***********"
           }
         ],
         "dryRun": "lookup",
         "endpoint": "https://*****.orange/api/v4",
         "token": "***********",
         "autodiscoverFilter": ["*****/otip/otip-helm"]
       }
```
**Scanned Helm chart.yaml**
```
---
annotations:
  category: EndUser
  licenses: Apache-2.0
apiVersion: v2
appVersion: 1.0.0
dependencies:
  - name: common
    repository: oci://registry-1.docker.io/bitnamicharts
    tags:
      - bitnami-common
    version: 2.0.0
    ...
 ```

### Relevant debug logs

<details><summary>Logs</summary>

```
 WARN: Host error (repository=****/otip-helm)
       "hostType": "docker",
       "packageName": "registry-1.docker.io/bitnamicharts/common",
       "err": {
         "name": "RequestError",
         "code": "EPROTO",
         "timings": {
           "start": 1693301850911,
           "socket": 1693301851065,
           "lookup": 1693301851065,
           "connect": 1693301851065,
           "upload": 1693301851163,
           "error": 1693301851164,
           "phases": {"wait": 154, "dns": 0, "tcp": 0, "request": 98, "total": 253}
         },
         "message": "write EPROTO C047C104EB7F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:355:\n",
         "stack": "RequestError: write EPROTO C047C104EB7F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:355:\n\n    at ClientRequest.<anonymous> (/opt/containerbase/tools/renovate/36.68.0/node_modules/got/dist/source/core/index.js:970:111)\n    at Object.onceWrapper (node:events:629:26)\n    at ClientRequest.emit (node:events:526:35)\n    at ClientRequest.emit (node:domain:489:12)\n    at ClientRequest.origin.emit (/opt/containerbase/tools/renovate/36.68.0/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (node:_http_client:501:9)\n    at TLSSocket.emit (node:events:526:35)\n    at TLSSocket.emit (node:domain:489:12)\n    at emitErrorNT (node:internal/streams/destroy:151:8)\n    at emitErrorCloseNT (node:internal/streams/destroy:116:3)\n    at processTicksAndRejections (node:internal/process/task_queues:82:21)\n    at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/36.68.0 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://registry-1.docker.io/v2/bitnamicharts/common/tags/list?n=10000",
           "hostType": "docker",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         }
       }
 INFO: External host error causing abort - skipping (repository=oln/nif/tools/otip/otip-helm)
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
Most likely you have a proxy (potentially invisible/MITM) which is breaking the SSL lookups. Do you have custom certificates in use?

**Reply by rcarre** [reply]
Thanks @rarkins for your swift reply. I would say there is no custom certificate when triggering docker.io ?

**Reply by rarkins** [reply]
Incorrect. If your network has a man in the middle proxy, then you need a custom certificate because it's trying to intercept your requests.

**Reply by rcarre** [reply]
I do catch what you mean. However, when placing an https request from the renovate container, I do not get ssl based error
`docker run -ti --env-file env.list -v "renovate":/usr/src/app -w /usr/src/app renovate/renovate:full curl -I "https://registry-1.docker.io/v2/bitnamicharts/common/tags/list?n=10000"`
```
HTTP/1.1 200 OK
Connection: close

HTTP/1.1 401 Unauthorized
content-type: application/json
docker-distribution-api-version: registry/2.0
www-authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="repository:bitnamicharts/common:pull"
date: Tue, 29 Aug 2023 11:05:50 GMT
content-length: 163
strict-transport-security: max-age=31536000
```

**Reply by rarkins** [reply]
Please do this:
- Create a reproduction repository with just one dependency which demonstrates the problem
- Run Renovate with LOG_LEVEL=debug, just on that one repository
- Redact the logs if you need, but be careful not to remove data which might be relevant!
- Upload the log to a gist
- Link to the gist here

**Reply by rcarre** [reply]
Thx for your help. Here you are: https://gist.github.com/rcarre/36ce7cc1fefcefb7d574472818b8786f

**Reply by rarkins** [reply]
Do you have any of the HTTP_PROXY variables in your env file or injected into the container some other way?

**Reply by rarkins** [reply]
Seems like it could be the same problem: https://github.com/axios/axios/issues/4840#issuecomment-1330970813

**Reply by rcarre** [reply]
> Do you have any of the HTTP_PROXY variables in your env file or injected into the container some other way?

yes I do from .docker/config.json
```
{
	"proxies": {
		"default": {
			"httpProxy": "http://172.17.0.1:3128/",
			"httpsProxy": "http://172.17.0.1:3128/",
			"noProxy": "localhost,127.0.0.0/8"
		}
	}
}
```

Same proxy is listening to 2 @IP:@Port
```
lo       127.0.0.1   3128
docker0  172.17.0.1  3128
```

**Reply by rarkins** [reply]
So that proxy is likely causing the problem. My guess:
- Renovate sends a https request to auth.docker.io
- The proxy intercepts the request, attempts to do a https handshake pretending to be auth.docker.io
- Renovate (or more specifically node.js) rejects the invalid certificate

If that proxy is going to do _interception_ of https then Renovate needs to be made aware of its certificates.

You _could_ set `NODE_TLS_REJECT_UNAUTHORIZED=0` in env but (a) that's very insecure, and (b) there may be failures later in non-Node components like maven, pip, etc.

**Reply by rcarre** [reply]
I get your point but in such a case request the following request should also fail, should not it ?
```
docker run -ti --env-file /home/mlpm8474/renovate/env.list -v "/home/mlpm8474/renovate":/usr/src/app -w /usr/src/app renovate/renovate:full curl -I "https://registry-1.docker.io/v2/bitnamicharts/common/tags/list?n=10000"
```

**Reply by rcarre** [reply]
Gave a try with `NODE_TLS_REJECT_UNAUTHORIZED=0` with same error
By the way, which env var is equivalent to `hostRules matchHost: 'docker.io' username: 'orangeoln`' ? `DOCKER_DOCKER_IO_USERNAME=orangeoln` or `DOCKER__DOCKER_IO_USERNAME=orangeoln`

**Reply by rarkins** [reply, marked as answer]
Curl possibly has special handling for this Docker proxy approach. A better comparison would be to write a plain Node.js https fetch script and test that

**Reply by rcarre** [reply]
You are right @rarkins,
thanks to Wireshark, I figured out TLS v1.3 handshake issues with [`alpaca` local HTTP proxy](https://github.com/samuong/alpaca)
I updated `.docker/config.json` to bypass alpaca local proxy and then target directly remote proxy ... everything is fine now.
Thanks for your help.

**Comment by rarkins** [comment]
FYI "insecure registry" means to query using http not https. It does not mean that you can allow SSL errors, so make sure it's removed

**Reply by rcarre** [reply]
Removal confirmed

**Comment by rcarre** [comment]
Fixed
